consensus/src/detail/mod.rs

```diff
@@ -206,6 +214,103 @@ impl Consensus {
+        self.make_ro_db_tx().db_tx.get_block_index(id).map_err(BlockError::from)
+    }
+
+    pub fn get_header_from_height(
+        &self,
+        height: &BlockHeight,
+    ) -> Result<Option<BlockHeader>, BlockError> {
+        let consensus_ref = self.make_ro_db_tx();
+        let id = consensus_ref
+            .db_tx
+            .get_block_id_by_height(height)?
+            .ok_or(BlockError::NotFound)?;
+        Ok(consensus_ref
+            .db_tx
+            .get_block_index(&id)?
+            .map(|block_index| block_index.get_block_header().clone()))
+    }
+
+    pub fn get_locator(&self) -> Result<Vec<BlockHeader>, BlockError> {
+        let id = self.get_best_block_id()?.ok_or(BlockError::NotFound)?;
+        let height = self.get_block_height_in_main_chain(&id)?.ok_or(BlockError::NotFound)?;
+
+        let headers = itertools::iterate(0, |&i| if i == 0 { 1 } else { i * 2 })
+            .take_while(|i| (height - BlockDistance::new(*i)).is_some())
+            .map(|i| {
+                self.get_header_from_height(
+                    &(height - BlockDistance::new(i)).expect("distance to be valid"),
+                )
+            });
+
+        itertools::process_results(headers, |iter| iter.flatten().collect::<Vec<_>>())
+    }
+
+    // TODO: use `ConsensusRef::is_block_in_main_chain()` implementation instead, how?
+    fn is_block_in_main_chain(&self, block_index: &BlockIndex) -> Result<bool, BlockError> {
+        let consensus_ref = self.make_ro_db_tx();
+        let height = block_index.get_block_height();
+        let id_at_height =
+            consensus_ref.db_tx.get_block_id_by_height(&height).map_err(BlockError::from)?;
+        match id_at_height {
+            Some(id) => Ok(id == *block_index.get_block_id()),
+            None => Ok(false),
+        }
+    }
+
+    pub fn get_headers(&self, locator: Vec<BlockHeader>) -> Result<Vec<BlockHeader>, BlockError> {
+        // use genesis block if no common ancestor with better block height is found
+        let mut best = BlockHeight::new(0);
+
+        for header in locator.iter() {
+            if let Some(block_index) = self.get_block_index(&header.get_id())? {
+                if self.is_block_in_main_chain(&block_index)? {
+                    best = block_index.get_block_height();
+                    break;
+                }
+            }
+        }
+
+        // get headers until either the best block or header limit is reached
+        let limit = std::cmp::min(
+            (best + HEADER_LIMIT).ok_or(BlockError::Unknown)?,
+            self.get_block_height_in_main_chain(
+                &self.get_best_block_id()?.expect("best block to exist"),
+            )?
+            .expect("best block's height to exist"),
+        );
+
+        let headers = itertools::iterate(best.next_height(), |iter| iter.next_height())
+            .take_while(|height| height <= &limit)
```

Comment:
I'm not entirely sure how to do this. Before I settled on using `process_results()` I was playing around with a normal iteration but that required to implement `std::iter::Step` for `BlockHeight` which is a nightly feature so I'm not certain if what you're asking for here is possible to do with a stable compiler. I'll leave this as unresolved for now.

Reply:
Feel free to resolve it if you don't want to deal with it, it's a tiny minor issue. I thought it would be possible to implement `Iterator` directly on the range but that might cause issues with overlapping trait impls so not sure.